rust/src/lib/ifaces/inter_ifaces.rs

```diff
@@ -281,6 +281,76 @@ impl Interfaces {
     }
 }
 
+pub(crate) struct MergedInterfaces<'a> {
+    inner: Vec<&'a Interface>,
+}
+
+impl<'a> MergedInterfaces<'a> {
+    pub(crate) fn merge(
+        current: &'a Interfaces,
+        add: &'a Interfaces,
+        update: &'a Interfaces,
+        delete: &'a Interfaces,
+    ) -> Self {
+        let mut merged = Vec::new();
+        for current in current.to_vec().into_iter() {
+            if delete
+                .get_iface(current.name(), current.iface_type())
+                .is_none()
+            {
+                merged.push(current);
+            }
+        }
+        for iface in merged.iter_mut() {
+            if let Some(updated) =
+                update.get_iface(iface.name(), iface.iface_type())
+            {
+                *iface = updated;
```

Comment:
When a bridge is only mentioned by its name with defining a port list, your code are treating it like losing all ports.

Reply:
And please place the code in `ifaces/inter_ifaces_controller.rs`

Reply:
In the merged state nothing is mentioned only by name. It keeps list of all interfaces so the port information should still be there, unless I am missing something. Also this code is not specific only for interfaces with controllers so I am not sure if the `ifaces/inter_ifaces_controller.rs` is the right place where to put it.

Reply:
The `update` is holding the interface mentioned in desire state and found in current state. It only contain information on user provides.

For example, you code does not fail on this yml. (Before this yml, br0 is holding eth1 as its port).

```yml
---
interfaces:
  - name: br0
  - name: br1
    type: linux-bridge
    state: up
    bridge:
      options:
        stp:
          enabled: false
      port:
        - name: eth1
```

Reply:
True but I am not sure if this can solved by the new code that is checking that. This issue is the same for other interfaces e.g. vlan. So we need to probably solve proper merging of current state and desired state somewhere in the `gen_state_for_apply`. Do you think it should be part of this PR?

Reply:
The vlan does not have overbook problem. You can just use current interface port list as fallback when desire is None.

For orphan vlan, iterate all interface to make sure vlan parent is found in current and not in absent list.

There is no other code need top level state merging. So please keep it dedicate for this purpose.

Reply:
Maybe I am still missing something, the merged state does not contain interfaces marked as absent. So the vlan cannot find it's parent that will be absent.

Reply:
Let me merge this PR and continue the work later.